File: src/native/libs/System.Native/pal_autoreleasepool.m
// Licensed to the .NET Foundation under one or more agreements.
// The .NET Foundation licenses this file to you under the MIT license.

#include "pal_autoreleasepool.h"
#include <Foundation/Foundation.h>
#include <objc/runtime.h>

#if __has_feature(objc_arc)
#error This file uses manual memory management and must not use ARC, but ARC is enabled.
#endif

void EnsureNSThreadIsMultiThreaded(void)
{
    if (![NSThread isMultiThreaded])
    {
        // Start another no-op thread with the NSThread APIs to get NSThread into multithreaded mode.
        // The NSAutoReleasePool APIs can't be used on secondary threads until NSThread is in multithreaded mode.
        // See https://developer.apple.com/documentation/foundation/nsautoreleasepool for more information.
        //
        // We need to use detachNewThreadSelector to put NSThread into multithreaded mode.
        // We can't use detachNewThreadWithBlock since it doesn't change NSThread into multithreaded mode for some reason.
        // See https://developer.apple.com/documentation/foundation/nswillbecomemultithreadednotification for more information.
        id placeholderObject = [[NSMutableString alloc] init];
        [NSThread detachNewThreadSelector:@selector(appendString:) toTarget:placeholderObject withObject:@""];
#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Wobjc-messaging-id"
        [placeholderObject release];
#pragma clang diagnostic pop
    }
    assert([NSThread isMultiThreaded]);
}

void* SystemNative_CreateAutoreleasePool(void)
{
    EnsureNSThreadIsMultiThreaded();
    return [[NSAutoreleasePool alloc] init];
}

void SystemNative_DrainAutoreleasePool(void* pool)
{
    [((NSAutoreleasePool*)pool) drain];
}

File: src/native/libs/System.Native/pal_autoreleasepool.c
// Licensed to the .NET Foundation under one or more agreements.
// The .NET Foundation licenses this file to you under the MIT license.

#include "pal_autoreleasepool.h"
#include <stdlib.h>
#include "pal_utilities.h"

#ifndef _MSC_VER
// Don't warning about not declaring a function with [[noreturn]] since it's only true in Debug mode.
#pragma GCC diagnostic ignored "-Wmissing-noreturn"
#endif

// These functions should not be used, but they need to be defined
// to satisfy the tooling we used to enable redirecting P/Invokes
// for the single file scenario.
void* SystemNative_CreateAutoreleasePool(void)
{
    assert_err(false, "Autorelease pools not supported on this platform.", EINVAL);
    return NULL;
}

void SystemNative_DrainAutoreleasePool(void* pool)
{
    (void)pool;
    assert_err(false, "Autorelease pools not supported on this platform.", EINVAL);
}
